From 76e05613ee28f4ac4a0ab97effc32e0e78e37a56 Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Thu, 29 Sep 2022 16:15:20 -0300
Subject: [PATCH] stdlib: Fix __getrandom_nocancel type and arc4random usage
 (BZ #29638)

Using an unsigned type prevents the fallback to be used if kernel
does not support getrandom syscall.

Checked on x86_64-linux-gnu.

Reviewed-by: Wilco Dijkstra  <Wilco.Dijkstra@arm.com>
(cherry picked from commit 13db9ee2cb3b77e25f852be7d6952882e1be6f00)
---
 NEWS                                 | 1 +
 stdlib/arc4random.c                  | 2 +-
 sysdeps/unix/sysv/linux/not-cancel.h | 2 +-
 3 files changed, 3 insertions(+), 2 deletions(-)

--- a/NEWS
+++ b/NEWS
@@ -29,6 +29,8 @@
   [29583] Use 64-bit interfaces in gconv_parseconfdir
   [29607] nscd repeatably crashes calling __strlen_avx2 when hosts cache is
     enabled
+  [29638] libc: stdlib: arc4random fallback is never used
+
 
 
 Version 2.36
--- a/stdlib/arc4random.c
+++ b/stdlib/arc4random.c
@@ -34,7 +34,7 @@
 __arc4random_buf (void *p, size_t n)
 {
   static int seen_initialized;
-  size_t l;
+  ssize_t l;
   int fd;
 
   if (n == 0)
--- a/sysdeps/unix/sysv/linux/not-cancel.h
+++ b/sysdeps/unix/sysv/linux/not-cancel.h
@@ -68,7 +68,7 @@
   INTERNAL_SYSCALL_CALL (writev, fd, iov, iovcnt);
 }
 
-static inline int
+static inline ssize_t
 __getrandom_nocancel (void *buf, size_t buflen, unsigned int flags)
 {
   return INLINE_SYSCALL_CALL (getrandom, buf, buflen, flags);
